import "maidsafe/dht/kademlia.proto";

package maidsafe.dht.protobuf;

message SignedValue {
  required bytes value = 1;
  required bytes signature = 2;
}

message PingRequest {
  required Contact sender = 1;
  required bytes ping = 2;
}

message PingResponse {
  required bytes echo = 1;
}

message FindValueRequest {
  required Contact sender = 1;
  required bytes key = 2;
  optional int32 num_nodes_requested = 3;
}

message FindValueResponse {
  required bool result = 1;
  repeated Contact closest_nodes = 2;
  repeated SignedValue signed_values = 3;
  optional Contact cached_copy_holder = 4;
}

message FindNodesRequest {
  required Contact sender = 1;
  required bytes key = 2;
  optional int32 num_nodes_requested = 3;
}

message FindNodesResponse {
  required bool result = 1;
  repeated Contact closest_nodes = 2;
}

message StoreRequest {
  required Contact sender = 1;
  required bytes key = 2;
  required SignedValue signed_value = 3;
  required int32 ttl = 4;
}

message StoreResponse {
  required bool result = 1;
}

message StoreRefreshRequest {
  required Contact sender = 1;
  optional bytes serialised_store_request = 2;
  optional bytes serialised_store_request_signature = 3;
}

message StoreRefreshResponse {
  required bool result = 1;
}

message DeleteRequest {
  required Contact sender = 1;
  required bytes key = 2;
  required SignedValue signed_value = 3;
}

message DeleteResponse {
  required bool result = 1;
}

message DeleteRefreshRequest {
  required Contact sender = 1;
  optional bytes serialised_delete_request = 2;
  optional bytes serialised_delete_request_signature = 3;
}

message DeleteRefreshResponse {
  required bool result = 1;
}

message DownlistNotification {
  required Contact sender = 1;
  repeated bytes node_ids = 2;
}
